### PR TITLE
enhance: optimize bitmap index

### PR DESCRIPTION
**Pull request by zhagnlu**
#32900

### CODE REVIEW

File: internal/core/src/index/HybridScalarIndex.cpp [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+    }
+    internal_index_type_ = InternalIndexType::NONE;
+}
+
+template <typename T>
+InternalIndexType
+HybridScalarIndex<T>::SelectIndexBuildType(size_t n, const T* values) {
+    std::set<T> distinct_vals;
+    for (size_t i = 0; i < n; i++) {
+        distinct_vals.insert(values[i]);
```
Comment (longjiquan) [inline internal/core/src/index/HybridScalarIndex.cpp:61]: break when distinct_vals.size() >= BITMAP_INDEX_CARDINALITY_BOUND

File: internal/core/src/index/HybridScalarIndex.cpp [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        if (distinct_vals.size() >= BITMAP_INDEX_CARDINALITY_BOUND) {
+            break;
+        }
+    }
+
+    // Decide whether to select bitmap index or marisa index
+    if (distinct_vals.size() >= BITMAP_INDEX_CARDINALITY_BOUND) {
+        internal_index_type_ = InternalIndexType::STLSORT;
```
Comment (longjiquan) [inline internal/core/src/index/HybridScalarIndex.cpp:87]: Should be marisa-trie index.

File: internal/core/src/index/HybridScalarIndex.cpp (hunk 1 above)
Comment (zhagnlu) [reply internal/core/src/index/HybridScalarIndex.cpp:61]: this interface is just used for test, test data common not big, not a serious problem. the real work interface do it

File: internal/core/src/index/HybridScalarIndex.cpp (hunk 2 above)
Comment (zhagnlu) [reply internal/core/src/index/HybridScalarIndex.cpp:87]: good catch

**Comment by zhagnlu** [comment]
rerun ut

File: internal/proxy/task_index.go [hunk 3]
```diff
@@ -354,6 +354,12 @@ func checkTrain(field *schemapb.FieldSchema, indexParams map[string]string) erro
 		return err
 	}
 
+	if indexType == indexparamcheck.IndexBitmap {
+		_, exist := indexParams[common.BitmapCardinalityLimitKey]
+		if !exist {
+			indexParams[common.BitmapCardinalityLimitKey] = paramtable.Get().CommonCfg.BitmapIndexCardinalityBound.GetValue()
```
Comment (czs007) [inline internal/proxy/task_index.go:360]: add indexParamChecker to check the range

**Comment by zhagnlu** [comment]
rerun ut

**Pull request merged**